File: app/src/main/java/com/ribic/nejc/veselica/data/PartyProvider.java
package com.ribic.nejc.veselica.data;

import android.content.ContentProvider;
import android.content.ContentValues;
import android.content.UriMatcher;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;


public class PartyProvider extends ContentProvider{

    public static final int CODE_PARTY = 100;
    public static final int CODE_PARTY_WITH_ID = 101;

    private static final UriMatcher sUriMatcher = buildUriMatcher();

    private PartyDbHelper mOpenHelper;

    public static UriMatcher buildUriMatcher(){
        final UriMatcher matcher = new UriMatcher(UriMatcher.NO_MATCH);
        final String authority = PartyContract.CONTENT_AUTHORITY;

        matcher.addURI(authority, PartyContract.PATH_PARTY, CODE_PARTY);
        matcher.addURI(authority, PartyContract.PATH_PARTY + "/#", CODE_PARTY_WITH_ID);
        return matcher;
    }




    @Override
    public boolean onCreate() {
        mOpenHelper = new PartyDbHelper(getContext());
        return true;
    }

    @Nullable
    @Override
    public Cursor query(@NonNull Uri uri, @Nullable String[] projection, @Nullable String selection, @Nullable String[] selectionArgs, @Nullable String sortOrder) {
        Cursor cursor;
        switch (sUriMatcher.match(uri)){
            case CODE_PARTY_WITH_ID: {
                String id = uri.getLastPathSegment();
                String[] selectionArgs2 = new String[]{id};
                cursor = mOpenHelper.getReadableDatabase().query(PartyContract.PartyEntry.TABLE_NAME,
                        projection,
                        PartyContract.PartyEntry.COLUMN_PARTY_ID + " = ? ",
                        selectionArgs2,
                        null,
                        null,
                        sortOrder);
                break;
            }
            case CODE_PARTY: {
                cursor = mOpenHelper.getReadableDatabase().query(
                        PartyContract.PartyEntry.TABLE_NAME,
                        projection,
                        selection,
                        selectionArgs,
                        null,
                        null,
                        sortOrder);
                break;
            }
            default:
                throw new UnsupportedOperationException("Unknown uri: " + uri);
        }

        return cursor;
    }

    @Nullable
    @Override
    public String getType(@NonNull Uri uri) {
        return null;
    }

    @Override
    public int bulkInsert(@NonNull Uri uri, @NonNull ContentValues[] values) {
        final SQLiteDatabase db = mOpenHelper.getWritableDatabase();
        switch (sUriMatcher.match(uri)){
            case CODE_PARTY:
                db.beginTransaction();
                int rowsInserted = 0;
                try{
                    for (ContentValues value : values){
                        long _id = db.insert(PartyContract.PartyEntry.TABLE_NAME, null, value);
                        if (_id != -1) rowsInserted++;
                    }
                    db.setTransactionSuccessful();
                } finally {
                    db.endTransaction();
                }
                if (rowsInserted > 0) getContext().getContentResolver().notifyChange(uri, null);
                return rowsInserted;
            default:
                return super.bulkInsert(uri, values);
        }
    }

    @Nullable
    @Override
    public Uri insert(@NonNull Uri uri, @Nullable ContentValues contentValues) {
        return null;
    }

    @Override
    public int delete(@NonNull Uri uri, @Nullable String selection, @Nullable String[] selectionArgs) {
        int numRowsDeleted;
        if (null == selection) selection = "1";
        switch (sUriMatcher.match(uri)){
            case CODE_PARTY:
                numRowsDeleted = mOpenHelper.getWritableDatabase().delete(
                        PartyContract.PartyEntry.TABLE_NAME,
                        selection,
                        selectionArgs);
                break;
            case CODE_PARTY_WITH_ID:
                String id = uri.getLastPathSegment();
                numRowsDeleted = mOpenHelper.getWritableDatabase().delete(
                        PartyContract.PartyEntry.TABLE_NAME,
                        "movie_id=?",
                        new String[]{id});
                break;
            default:
                throw new UnsupportedOperationException("Unknown uri: " + uri);
        }
        if (numRowsDeleted != 0){
            getContext().getContentResolver().notifyChange(uri, null);
        }
        return numRowsDeleted;
    }

    @Override
    public int update(@NonNull Uri uri, @Nullable ContentValues contentValues, @Nullable String s, @Nullable String[] strings) {
        return 0;
    }

    @Override
    public void shutdown() {
        mOpenHelper.close();
        super.shutdown();
    }
}

File: app/src/main/java/com/ribic/nejc/veselica/data/PartyProvider.kt
package com.ribic.nejc.veselica.data

import android.content.ContentProvider
import android.content.ContentValues
import android.content.UriMatcher
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.net.Uri


class PartyProvider : ContentProvider() {

    private var mOpenHelper: PartyDbHelper? = null


    override fun onCreate(): Boolean {
        mOpenHelper = PartyDbHelper(context)
        return true
    }

    override fun query(uri: Uri, projection: Array<String>?, selection: String?, selectionArgs: Array<String>?, sortOrder: String?): Cursor? {
        val cursor: Cursor
        when (sUriMatcher.match(uri)) {
            CODE_PARTY_WITH_ID -> {
                val id = uri.lastPathSegment
                val selectionArgs2 = arrayOf(id)
                cursor = mOpenHelper!!.readableDatabase.query(PartyContract.PartyEntry.TABLE_NAME,
                        projection,
                        PartyContract.PartyEntry.COLUMN_PARTY_ID + " = ? ",
                        selectionArgs2, null, null,
                        sortOrder)
            }
            CODE_PARTY -> {
                cursor = mOpenHelper!!.readableDatabase.query(
                        PartyContract.PartyEntry.TABLE_NAME,
                        projection,
                        selection,
                        selectionArgs, null, null,
                        sortOrder)
            }
            else -> throw UnsupportedOperationException("Unknown uri: " + uri)
        }

        return cursor
    }

    override fun getType(uri: Uri): String? {
        return null
    }

    override fun bulkInsert(uri: Uri, values: Array<ContentValues>): Int {
        val db = mOpenHelper!!.writableDatabase
        when (sUriMatcher.match(uri)) {
            CODE_PARTY -> {
                db.beginTransaction()
                var rowsInserted = 0
                try {
                    for (value in values) {
                        val _id = db.insert(PartyContract.PartyEntry.TABLE_NAME, null, value)
                        if (_id.toInt() != -1) rowsInserted++
                    }
                    db.setTransactionSuccessful()
                } finally {
                    db.endTransaction()
                }
                if (rowsInserted > 0) context!!.contentResolver.notifyChange(uri, null)
                return rowsInserted
            }
            else -> return super.bulkInsert(uri, values)
        }
    }

    override fun insert(uri: Uri, contentValues: ContentValues?): Uri? {
        return null
    }

    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?): Int {
        var selection = selection
        val numRowsDeleted: Int
        if (null == selection) selection = "1"
        when (sUriMatcher.match(uri)) {
            CODE_PARTY -> numRowsDeleted = mOpenHelper!!.writableDatabase.delete(
                    PartyContract.PartyEntry.TABLE_NAME,
                    selection,
                    selectionArgs)
            CODE_PARTY_WITH_ID -> {
                val id = uri.lastPathSegment
                numRowsDeleted = mOpenHelper!!.writableDatabase.delete(
                        PartyContract.PartyEntry.TABLE_NAME,
                        "movie_id=?",
                        arrayOf(id))
            }
            else -> throw UnsupportedOperationException("Unknown uri: " + uri)
        }
        if (numRowsDeleted != 0) {
            context!!.contentResolver.notifyChange(uri, null)
        }
        return numRowsDeleted
    }

    override fun update(uri: Uri, contentValues: ContentValues?, s: String?, strings: Array<String>?): Int {
        return 0
    }

    override fun shutdown() {
        mOpenHelper!!.close()
        super.shutdown()
    }

    companion object {

        val CODE_PARTY = 100
        val CODE_PARTY_WITH_ID = 101

        private val sUriMatcher = buildUriMatcher()

        fun buildUriMatcher(): UriMatcher {
            val matcher = UriMatcher(UriMatcher.NO_MATCH)
            val authority = PartyContract.CONTENT_AUTHORITY

            matcher.addURI(authority, PartyContract.PATH_PARTY, CODE_PARTY)
            matcher.addURI(authority, PartyContract.PATH_PARTY + "/#", CODE_PARTY_WITH_ID)
            return matcher
        }
    }
}
